apps/bf_prefix.c

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 1995-2016 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+#include <openssl/bio.h>
+#include "apps.h"
+
+static int prefix_write(BIO *b, const char *out, size_t outl,
+                        size_t *numwritten);
+static int prefix_read(BIO *b, char *buf, size_t size, size_t *numread);
+static int prefix_puts(BIO *b, const char *str);
+static int prefix_gets(BIO *b, char *str, int size);
+static long prefix_ctrl(BIO *b, int cmd, long arg1, void *arg2);
+static int prefix_destroy(BIO *b);
+static long prefix_callback_ctrl(BIO *b, int cmd, BIO_info_cb *fp);
+
+/* #define DEBUG */
+
+static BIO_METHOD *prefix_meth = NULL;
+
+BIO_METHOD *apps_bf_prefix(void)
+{
+    if (prefix_meth == NULL) {
+        if ((prefix_meth =
+             BIO_meth_new(BIO_TYPE_FILTER, "Prefix filter")) == NULL
+            || !BIO_meth_set_destroy(prefix_meth, prefix_destroy)
+            || !BIO_meth_set_write_ex(prefix_meth, prefix_write)
+            || !BIO_meth_set_read_ex(prefix_meth, prefix_read)
+            || !BIO_meth_set_puts(prefix_meth, prefix_puts)
+            || !BIO_meth_set_gets(prefix_meth, prefix_gets)
+            || !BIO_meth_set_ctrl(prefix_meth, prefix_ctrl)
+            || !BIO_meth_set_callback_ctrl(prefix_meth, prefix_callback_ctrl)) {
+            BIO_meth_free(prefix_meth);
+            prefix_meth = NULL;
+        }
+    }
+    return prefix_meth;
+}
+
+static int prefix_destroy(BIO *b)
+{
+    OPENSSL_free(BIO_get_data(b));
+    return 1;
+}
+
+static int prefix_read(BIO *b, char *in, size_t size, size_t *numread)
+{
+    return BIO_read_ex(BIO_next(b), in, size, numread);
+}
+
+static int prefix_write(BIO *b, const char *out, size_t outl,
+                        size_t *numwritten)
+{
+    const char *prefix = (const char *)BIO_get_data(b);
+
+    *numwritten = 0;
+
+    if (BIO_number_written(b) == 0) {
+        size_t dontcare;
```

Comment:
If you moved all the variable decls up to the top, it would save some lines.  That's generally how we code, even if it's not a style.

Reply:
> That's generally how we code, even if it's not a style.

Is it? In general, I think keeping variables local to a block is a good thing.

Reply:
Look at everything else, at least in the apps directory. Everything is declared and init'd up at top.

Reply:
There are many examples in the whole codebase where we do this kind of thing. I've not checked apps in detail but in any case I'm not convinced the code in apps is what we should be setting our benchmark against. AFAIK we have never had a rule that prevents us declaring vars in this way, and I don't think we should start now.

Reply:
Most of our code declares everything at the top and then does goto err.  I am surprised to see you disagreeing with that. I was mentioned apps because I am most familiar with it, and this is where the code is being changed.

If left this way, I'm not going to -1 it, but I will also leave it to someone else to approve.

Reply:
Well, historically, it's true that everything was cramed at the top.  I strongly dislike doing so for variables that are really only used inside one block, like this one, and I'm trying hard to get us into this habit by example.  So no, I don't see a reason to move this particular variable unless it becomes useful elsewhere.

Reply:
> Most of our code declares everything at the top and then does goto err.

That is one particular pattern and it is not mutually exclusive with declaring local block variables. Obviously variables freed in the err section need to be visible to the whole function - but not other variables. In this particular case there is no usage of the "goto err" pattern, so I fail to see its relevance.

Anyway I'll leave it to @levitte to decide the most appropriate action in this particular case. But I would hate to see this particular "rule" becoming a de facto standard.

Reply:
The BSD KNF/style(9) is frequently misread to imply that all locals must be declared at the top of the function scope.  There was some recent discussion on one of the FreeBSD lists about how this is not actually the case, and declaring variables within inner scope is compliant with style(9) especially in long functions where it would otherwise be confusing.

Reply:
Thanks. I wasn't misreading.  I was saying I would like this to be like most of the rest of our code.

Reply:
@kaduk:
> ... especially in long functions where it would otherwise be confusing.

This, exactly this is my main reason to want us to get into a better habit.  I know that's this particular function isn't exactly huge, so that isn't a worry, but I think that getting into a better habit (and I do think it's better) takes some training, yeah?